mlx/fast.cpp

```diff
@@ -127,4 +127,168 @@ bool RoPE::is_equivalent(const Primitive& other) const {
       offset_ == a_other.offset_);
 }
 
+/** Computes: O = softmax(Q @ K.T) @ V **/
+array scaled_dot_product_attention(
+    const array& queries,
+    const array& keys,
+    const array& values,
+    const float scale,
+    const std::optional<array>& mask,
+    StreamOrDevice s) {
+  for (const auto& tensor : {queries, keys, values}) {
+    if (tensor.ndim() != 4) {
+      std::ostringstream msg;
+      msg << "[scaled_dot_product_attention] input with shape "
+          << tensor.shape() << " expected to be rank 4";
+      throw std::invalid_argument(msg.str());
+    }
+  }
+
+  size_t batch_dim = queries.shape(0);
+  if (batch_dim != 1) {
+    std::ostringstream msg;
+    msg << "[scaled_dot_product_attention] Expected batch dimension of 1. ";
+    throw std::invalid_argument(msg.str());
+  }
```

Comment:
Correct me if I'm wrong, but it looks like the fallback will support batch > 1. If that is correct (which would be great), then maybe we use this in the dispatch condition rather than throwing here?

Reply:
yes, let's do that